extensions/gdata/src/com/google/refine/extension/gdata/GoogleAPIExtension.java

```diff
@@ -0,0 +1,205 @@
+
+    private static final String[] SCOPES = {DriveScopes.DRIVE, SheetsScopes.SPREADSHEETS, FusiontablesScopes.FUSIONTABLES};
+    
+    static public String getAuthorizationUrl(ButterflyModule module, HttpServletRequest request)
+            throws MalformedURLException {
+        String authorizedUrl = makeRedirectUrl(module, request);
+              
+        
+        GoogleAuthorizationCodeRequestUrl url = new GoogleAuthorizationCodeRequestUrl(
+                CLIENT_ID, 
+                authorizedUrl, // execution continues at authorized on redirect
+                Arrays.asList(SCOPES));
+        
+        return url.toString();
+
+    }
+
+    private static String makeRedirectUrl(ButterflyModule module, HttpServletRequest request)
+            throws MalformedURLException {
+        StringBuffer sb = new StringBuffer(module.getMountPoint().getMountPoint());
+        sb.append("authorized?winname=");
+        sb.append(ParsingUtilities.encode(request.getParameter("winname")));
+        sb.append("&cb=");
+        sb.append(ParsingUtilities.encode(request.getParameter("cb")));
+    
+        URL thisUrl = new URL(request.getRequestURL().toString());
+        URL authorizedUrl = new URL(thisUrl, sb.toString());
+
+        return authorizedUrl.toExternalForm();
+    }
+
+    static public String getTokenFromCode(ButterflyModule module, HttpServletRequest request) 
+            throws IOException {
+        String redirectUrl = makeRedirectUrl(module, request);
+        StringBuffer fullUrlBuf = request.getRequestURL();
+        if (request.getQueryString() != null) {
+          fullUrlBuf.append('?').append(request.getQueryString());
+        }
+        AuthorizationCodeResponseUrl authResponse =
+            new AuthorizationCodeResponseUrl(fullUrlBuf.toString());
+        // check for user-denied error
+        if (authResponse.getError() != null) {
+          // authorization denied...
+        } else {
+          // request access token using authResponse.getCode()...
+            String code = authResponse.getCode();
+            GoogleTokenResponse response = new GoogleAuthorizationCodeTokenRequest(HTTP_TRANSPORT,
+                    JSON_FACTORY, CLIENT_ID, CLIENT_SECRET, code, redirectUrl).execute();
+            String tokenAndExpiresInSeconds = response.getAccessToken() + "," + response.getExpiresInSeconds();
+            return tokenAndExpiresInSeconds;
+        }
+        return null;
+      }
+    
+    static public Drive getDriveService(String token) {
+        GoogleCredential credential = new GoogleCredential().setAccessToken(token);
+        
+        return new Drive.Builder(HTTP_TRANSPORT, JSON_FACTORY, credential).setHttpRequestInitializer(new HttpRequestInitializer() {
+            @Override
+            public void initialize(HttpRequest httpRequest) throws IOException {
+                credential.initialize(httpRequest);
+                httpRequest.setConnectTimeout(300 * 60000);  // 300 minutes connect timeout
```

Comment:
@jackyq2015 Is this timeout value still what Google recommends ?  I know that Apache HTTP has a internal default of 3 minutes timeout that can be overridden in Apache's newer package version.

Reply:
well, it is not recommendation from Google. 300 mins seems too much, Let me find out what value is better.

Reply:
set to 3 mins. it make more sense.